include/ginkgo/core/config/config.hpp

```diff
@@ -0,0 +1,175 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+
+
+#include <map>
+#include <string>
+#include <unordered_map>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/config/type_descriptor.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+
+namespace gko {
+namespace config {
+
+class registry;
+
+
+class pnode;
+
+
+/**
+ * parse is the main entry point to create an Ginkgo LinOpFactory based on
+ * some file configuration. It reads a configuration stored as a property tree
+ * and creates the desired type.
+ *
+ * General rules for configuration
+ * 1. The configuration can be used to define factory parameters and class
+ *    template parameters. Any factory parameter that is not defined in the
+ *    configuration will fallback to their default value. Any template parameter
+ *    that is not defined will fallback to the type_descriptor argument
+ * 2. The new `"type"` key determines which Ginkgo object to create. The value
+ *    for this key is the desired class name with namespaces (except for
+ *    `gko::`, `experimental::`). Any template parameters a class might have are
+ *    left out. Only classes with a factory are supported. For example, the
+ *    configuration `"type": "solver::Cg"` specifies that a Cg solver will be
+ *    created. Note: template parameters can either be given in the
+ *    configuration as separate key-value pairs, or in the type_descriptor.
+ * 3. Factory and class template parameters can be defined with key-value pairs
+ *    that are derived from the class they are refering to. When a factory has a
+ *    parameter with the function `with_<key>(value)`, then the configuration
+ *    allows `"<key>": value` to define this parameter. When a class has a
+ *    template parameter `template<typename/type... key> class`, then the
```

Comment:
```suggestion
 *    template parameter `template<typename/class... key> class`, then the
```
I don't think `type` is valid for a template parameter.

Reply:
`type` here is to present something like value template `isai_enum isai_type` or `bool ...`

Reply:
right. I'm not sure how clear that is, but it's relatively minor and could be fixed later.

Reply:
I use specific_type now, hope it is more clear